pallets/connectors-gateway/routers/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+		);
+
+		Ok(())
+	}
+
+	/// NOTE - the sender account ID provided here will be converted to an EVM
+	/// address via truncating. When the call is processed by the underlying EVM
+	/// pallet, this EVM address will be converted back into a substrate account
+	/// which will be charged for the transaction. This converted substrate
+	/// account is not the same as the original account.
+	pub fn do_send(&self, sender: T::AccountId, msg: Vec<u8>) -> DispatchResult {
+		let sender_evm_address = H160::from_slice(&sender.as_ref()[0..20]);
+
+		// TODO(cdamian): This returns a `DispatchResultWithPostInfo`. Should we
+		// propagate that to another layer that will eventually charge for the
+		// weight in the PostDispatchInfo?
+		<pallet_ethereum_transaction::Pallet<T> as EthereumTransactor>::call(
+			sender_evm_address,
+			self.evm_domain.target_contract_address,
+			msg.as_slice(),
+			self.evm_domain.fee_values.value,
+			self.evm_domain.fee_values.gas_price,
+			self.evm_domain.fee_values.gas_limit,
+		)
+		.map_err(|e| e.error)?;
+
+		Ok(())
+	}
+}
+
+/// The EVMDomain holds all relevant information for validating and executing
+/// the EVM call.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct EVMDomain {
+	/// The address of the contract deployed in our EVM.
+	pub target_contract_address: H160,
+
+	/// The `BlakeTwo256` hash of the target contract code.
+	///
+	/// This is used during router initialization to ensure that the correct
+	/// contract code is used.
+	pub target_contract_hash: H256,
+
+	/// The values used when executing the EVM call.
+	pub fee_values: FeeValues,
+}
+
+/// The FeeValues holds all information related to the transaction costs.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct FeeValues {
+	/// The value used when executing the EVM call.
+	pub value: U256,
+
+	/// The gas price used when executing the EVM call.
+	pub gas_price: U256,
+
+	/// The gas limit used when executing the EVM call.
+	pub gas_limit: U256,
+}
+
+/// A generic router used for executing XCM calls.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct XCMRouter<T>
+where
+	T: frame_system::Config + pallet_xcm_transactor::Config,
+{
+	pub xcm_domain: XcmDomain<T::CurrencyId>,
+	pub _marker: PhantomData<T>,
+}
+
+impl<T> XCMRouter<T>
+where
+	T: frame_system::Config + pallet_xcm_transactor::Config,
+{
+	/// Sets the weight information for the provided XCM domain location, and
+	/// the fee per second for the provided fee asset location.
+	pub fn do_init(&self) -> DispatchResult {
+		pallet_xcm_transactor::Pallet::<T>::set_transact_info(
+			<T as frame_system::Config>::RuntimeOrigin::root(),
+			self.xcm_domain.location.clone(),
+			self.xcm_domain.transact_info.transact_extra_weight,
+			self.xcm_domain.transact_info.max_weight,
+			self.xcm_domain.transact_info.transact_extra_weight_signed,
+		)?;
```

Comment:
Could we run into troubles if 

* `RouterA` sets these values
* time passes and we add a new `RouterB`
* `RouterB` overwrites values with a new value?

Reply:
I don't think so. In theory, we should only have one router for a specific location, so if at some point we overwrite the router I think it makes sense to overwrite the xcm transactor info that is stored.